include/oneapi/dpl/internal/dynamic_selection_impl/dynamic_load_policy_impl.h

```diff
@@ -0,0 +1,164 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_DYNAMIC_LOAD_POLICY_IMPL_H
+#define _ONEDPL_DYNAMIC_LOAD_POLICY_IMPL_H
+
+#include <atomic>
+#include <memory>
+#include <ostream>
+#include <vector>
+#include "oneapi/dpl/internal/dynamic_selection_impl/scoring_policy_defs.h"
+
+namespace oneapi {
+namespace dpl{
+namespace experimental{
+
+  template <typename Scheduler>
+  struct dynamic_load_policy_impl {
+    using scheduler_t = Scheduler;
+    using native_resource_t = typename scheduler_t::native_resource_t;
+    using execution_resource_t = typename scheduler_t::execution_resource_t;
+    using native_sync_t = typename scheduler_t::native_sync_t;
+    using load_t = int;
+
+    struct resource_t {
+      execution_resource_t e_;
+      std::atomic<load_t> load_;
+      resource_t(execution_resource_t e) : e_(e) { load_ = 0; }
+    };
+
+    using universe_container_t = std::vector<std::shared_ptr<resource_t>>;
+    using universe_container_size_t = typename universe_container_t::size_type;
+
+    struct dl_property_handle_t {
+      std::shared_ptr<resource_t> resource_ptr_;
+      static constexpr bool should_report_task_submission = true;
+      static constexpr bool should_report_task_completion = true;
+      static constexpr bool should_report_task_execution_time = false;
+
+      void report(const oneapi::dpl::experimental::property::task_submission_t&) const noexcept {
+        auto v = resource_ptr_->load_.fetch_add(1);
+      }
+      void report(const oneapi::dpl::experimental::property::task_completion_t&)  const noexcept {
+        auto v = resource_ptr_->load_.fetch_sub(1);
+      }
+      dl_property_handle_t(std::shared_ptr<resource_t> r) : resource_ptr_(r) {}
+    };
+
+    struct selection_handle_t {
+      using property_handle_t = dl_property_handle_t;
+      using native_resource_t = typename execution_resource_t::native_resource_t;
+
+      property_handle_t h_;
+
+      selection_handle_t(property_handle_t h) : h_(h) {}
+      native_resource_t get_native() { return h_.resource_ptr_->e_.get_native(); }
+      auto get_property_handle() { return h_; }
+    };
+
+    std::shared_ptr<scheduler_t> sched_;
+    universe_container_t universe_;
```

Comment:
Need to have both sched_ and universe_ accessed via shared_ptr to support reference semantics. And also should like make private things private instead of struct with all public.